jax_distribution/systems/ff_ippo_rware.py

```diff
@@ -389,55 +512,169 @@ def learner_setup(env: Environment, config: Dict) -> Tuple[callable, RunnerState
+    env = RwareMultiAgentWrapper(env)
+    env = AutoResetWrapper(env)
+    env = LogWrapper(env)
+    eval_env = jumanji.make(config["ENV_NAME"])
+    eval_env = RwareMultiAgentWrapper(eval_env)
+
+    # PRNG keys.
+    rng, rng_e, rng_p = jax.random.split(jax.random.PRNGKey(config["SEED"]), num=3)
     # Setup learner.
-    learn, initial_runner_state = learner_setup(env, config)
+    learn, runner_state = learner_setup(env, (rng, rng_p), config)
+
+    # Setup evaluator.
+    evaluator, absolute_metric_evaluator, (trained_params, eval_rngs) = evaluator_setup(
+        eval_env, rng_e, runner_state[0], config
+    )
 
     # Calculate total timesteps.
     n_devices = len(jax.devices())
-    total_timesteps = (
+    config["NUM_UPDATES_PER_EVAL"] = config["NUM_UPDATES"] // config["NUM_EVALUATION"]
+    timesteps_per_training = (
         n_devices
-        * config["NUM_UPDATES"]
+        * config["NUM_UPDATES_PER_EVAL"]
         * config["ROLLOUT_LENGTH"]
         * config["UPDATE_BATCH_SIZE"]
         * config["NUM_ENVS"]
     )
 
-    # Run experiment.
+    # Compile learner and evaluator.
     with TimeIt(tag="COMPILATION"):
-        learn(initial_runner_state)
-
-    with TimeIt(tag="EXECUTION", environment_steps=total_timesteps):
-        learner_output = learn(initial_runner_state)
-
-    return learner_output
+        learn(runner_state)
+    with TimeIt(tag="COMPILATION"):
+        evaluator(trained_params, eval_rngs)
+
+    # Run experiment for a total number of evaluations.
+    max_episode_return = jnp.float32(0.0)
+    best_params = None
+    for i in range(config["NUM_EVALUATION"]):
+        # Train.
+        with TimeIt(tag="EXECUTION", environment_steps=timesteps_per_training):
+            learner_output = learn(runner_state)
+            jax.block_until_ready(learner_output)
+
+        # Prepare for evaluation.
+        trained_params = jax.tree_util.tree_map(
+            lambda x: x[:, 0, ...], learner_output["runner_state"][0]
+        )
+        rng_e, *eval_rngs = jax.random.split(rng_e, n_devices + 1)
+        eval_rngs = jnp.stack(eval_rngs)
+        eval_rngs = eval_rngs.reshape(n_devices, -1)
+
+        # Evaluate.
+        evaluator_output = evaluator(trained_params, eval_rngs)
+        jax.block_until_ready(evaluator_output)
+
+        # Log the results
+        log(
+            metrics=learner_output,
+            t_env=timesteps_per_training * (i + 1),
+            trainer_metric=True,
+        )
+        episode_return = log(
+            metrics=evaluator_output["metrics"],
+            t_env=timesteps_per_training * (i + 1),
+        )
+        if config["ABSOLUTE_METRIC"] and max_episode_return <= episode_return:
+            best_params = copy.deepcopy(trained_params)
+            max_episode_return = episode_return
+
+        # Update runner state to continue training.
+        runner_state = learner_output["runner_state"]
+
+    if config["ABSOLUTE_METRIC"]:
```

Comment:
I don't think this should even be a config option, we should just do it. If it must be a config option I think "final_eval" is a better name

Reply:
the absolute metric does exist in theory with this name as it is a long evaluation but with the best trained network, also we need to keep it optional as we just use it if we need to produce eval_marl plots only